engine/socket.go

```diff
@@ -294,11 +294,14 @@ func (s *socket) setTransport(transport transports.Transport) {
 // Upon transport "drain" event
 func (s *socket) onDrain() {
 	if seqFn, err := s.sentCallbackFn.Shift(); err == nil {
-		socket_log.Debug("executing batch send callback")
+		socket_log.Debug("executing batch send callback with %d callbacks", len(seqFn))
 		for _, fn := range seqFn {
 			fn(s.Transport())
 		}
 	}
+
+	// Ensure any buffered packets are sent after transport becomes writable again
+	s.flush()
```

Comment:
Have you verified that adding `s.flush()` inside `onDrain()` will not cause a deadlock?

Reply:
Indeed, you are right about that issue. I've made two optimizations:

I now call s.flush() within a goroutine in onDrain.
I removed the premature triggering of the drain event within flush, because s.Transport().Send(wbuf) sends data via a goroutine, and also triggers the drain event after sending. Triggering the drain event too early could lead to closeTransport being called before all data has been sent in Close.